File: QuantumRand.py
#******************************************
#Title: Quantum Random
#Author: Aleksandar Damnjanovic
#Date: 19.11.2021
#Last edit: 20.11.2021
#******************************************

import requests
import json
import time
import sys

class QuantumRand:

    def __init__(self):
        self.next()

    def next(self):
        __address= 'https://qrng.anu.edu.au/API/jsonI.php?length=1&type=hex16&size=8'
        __data=requests.get(__address)
        __obj= json.loads(__data.text)
        __hexString= __obj['data'][0]
        self.value= int(__hexString,16)
        time.sleep(0.5)

    def getRand(self, min=0, max=sys.maxsize):
        if min<0 or max>sys.maxsize:
            print(f'min must be bigger than or 0\nmax must be less that {sys.maxsize}')

        __ret=0
        if max!=sys.maxsize:
            if self.value > max:
                __ret= self.value % max

        if __ret<min and min!=0:
            __ret= min + (self.value % (max-min))

        return __ret

    def getFloat(self, decimal=2):
        if self.value<sys.maxsize:
            __ret= float(self.value/ sys.maxsize)
        else:
            __ret= float(sys.maxsize/self.value)

        return f'%.{decimal}f' % __ret

File: QuantumRand.R
#******************************************
#Title: Quantum Random
#Author: Aleksandar Damnjanovic
#Date: 19.11.2021
#Last edit: 20.11.2021
#******************************************

library("httr") 
library("jsonlite")

nextOne<-function(){
  addr<-GET("https://qrng.anu.edu.au/API/jsonI.php?length=1&type=hex16&size=4")
  page<-content(addr,"text")
  json<-fromJSON(page, flatten = TRUE)
  json<-as.data.frame(json)
  val<-json$data
  val<-paste0('0x', val)
  val<-as.numeric(val)
  val<-sprintf("%.0f", val)
  Sys.sleep(.5)
  return(val)
}

getRand<-function(minv, maxv){
  bigVal<-nextOne()
  bigVal<-as.numeric(bigVal)
  minv<-as.numeric(minv)
  maxv<-as.numeric(maxv)
  
  if(minv<0 || maxv>.Machine$integer.max)
    print(paste('min must be bigger than or0;', 'max must be less than', .Machine$integer.max))

  ret<-0
  con<-(bigVal>maxv)
  if(maxv!=.Machine$integer.max)
    if(con)
      ret<-bigVal %% maxv
  
  if(ret<minv && minv!=0)
    ret= minv + (bigVal %% (maxv-minv))
  
  ret<-sprintf("%.0f", ret)
  return(ret)  
}

getFloat<-function(decimal){
  ret<-getRand(0,1000)
  ret<-as.numeric(ret)
  ret<- as.double(ret/1000)
  f<-paste0('%.',decimal,'f')
  ret<-sprintf(f, ret)
  return(ret)
}